### PR TITLE
Separate make targets that require elevated privileges

### PR DESCRIPTION
**Pull request by chofnar**
**What type of PR is this?**
cleanup

**Which issue does this PR fix**:
#459 

**What does this PR do / Why do we need it**:
This PR separates the tests that need some sort of privileged execution from the ones that don't by use of new make targets.
This allows running without needing privileges or when they are not available.
Additionally, this PR will now make tests run on every platform (make test was only run on linux && amd64).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/storage/storage_fs_sudo_test.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+		content = []byte("test-data3")
+		buf = bytes.NewBuffer(content)
+		buflen = buf.Len()
+		digest = godigest.FromBytes(content)
+		blob, err = imgStore.PutBlobChunkStreamed("dedupe2", upload, buf)
+		So(err, ShouldBeNil)
+		So(blob, ShouldEqual, buflen)
+
+		cmd := exec.Command("sudo", "chattr", "+i", path.Join(dir, "dedupe2", "blobs/sha256", blobDigest1)) // nolint: gosec
```
Comment (andaaron) [inline pkg/storage/storage_fs_sudo_test.go:77]: Do we still need the sudo command?

File: Makefile [hunk 2]
```diff
@@ -47,16 +47,26 @@ exporter-minimal: modcheck
...
+test: check-skopeo test/data $(NOTATION)
 	go test -tags extended,containers_image_openpgp -v -trimpath -race -timeout 15m -cover -coverpkg ./... -coverprofile=coverage-extended.txt -covermode=atomic ./...
 	go test -tags minimal,containers_image_openpgp -v -trimpath -race -cover -coverpkg ./... -coverprofile=coverage-minimal.txt -covermode=atomic ./...
 	# development-mode unit tests possibly using failure injection
 	go test -tags dev,extended,containers_image_openpgp -v -trimpath -race -timeout 15m -cover -coverpkg ./... -coverprofile=coverage-dev-extended.txt -covermode=atomic ./pkg/test/... ./pkg/storage/... ./pkg/extensions/sync/... -run ^TestInject
 	go test -tags dev,minimal,containers_image_openpgp -v -trimpath -race -cover -coverpkg ./... -coverprofile=coverage-dev-minimal.txt -covermode=atomic ./pkg/test/... ./pkg/storage/... ./pkg/extensions/sync/... -run ^TestInject
 
+.PHONY: privileged-test
+privileged-test: check-skopeo certs.d $(NOTATION)
+	go test -tags needsudo,extended,containers_image_openpgp -v -trimpath -race -timeout 15m -cover -coverpkg ./... -coverprofile=coverage-dev-needsudo.txt -covermode=atomic ./pkg/storage/... ./pkg/cli/... -run ^TestSudo
```
[the remark this answers is not in the document]
Comment (chofnar) [reply Makefile]: Renamed tests requiring privileges to start with "TestElevatedPrivileges"

File: Makefile [hunk 3]
```diff
@@ -81,7 +91,7 @@ $(NOTATION):
 .PHONY: covhtml
 covhtml:
 	go install github.com/wadey/gocovmerge@latest
-	gocovmerge coverage-minimal.txt coverage-extended.txt coverage-dev-minimal.txt coverage-dev-extended.txt > coverage.txt
+	gocovmerge coverage-minimal.txt coverage-extended.txt coverage-dev-minimal.txt coverage-dev-extended.txt coverage-dev-needsudo.txt > coverage.txt
```
[the remark this answers is not in the document]
Comment (chofnar) [reply Makefile]: Changed to coverage*.txt instead of explicit

File: Makefile [hunk 4]
```diff
@@ -47,16 +47,26 @@ exporter-minimal: modcheck
...
+.PHONY: privileged-test
+privileged-test: check-skopeo certs.d $(NOTATION)
+	go test -tags needsudo,extended,containers_image_openpgp -v -trimpath -race -timeout 15m -cover -coverpkg ./... -coverprofile=coverage-dev-needsudo.txt -covermode=atomic ./pkg/storage/... ./pkg/cli/... -run ^TestSudo
+
+test/data: check-skopeo
+	$(shell mkdir -p test/data; cd test/data; ../scripts/gen_certs.sh; cd ${TOP_LEVEL}; skopeo --insecure-policy copy -q docker://public.ecr.aws/t0x7q1g8/centos:7 oci:${TOP_LEVEL}/test/data/zot-test:0.0.1;skopeo --insecure-policy copy -q docker://public.ecr.aws/t0x7q1g8/centos:8 oci:${TOP_LEVEL}/test/data/zot-cve-test:0.0.1)
+	$(shell chmod -R a=rwx test/data)
+
+.PHONY: certs.d
+certs.d: test/data
```
[the remark this answers is not in the document]
Comment (chofnar) [reply Makefile]: Moved logic inside the specific Go test, and removed obsolete make targets

File: .github/workflows/ci-cd.yml [hunk 5]
```diff
@@ -58,9 +58,11 @@ jobs:
           echo "Building for $OS:$ARCH"
           cd $GITHUB_WORKSPACE
           if [[ $OS == "linux" && $ARCH == "amd64" ]]; then
-            make OS=$OS ARCH=$ARCH
+            make OS=$OS ARCH=$ARCH modcheck swagger binary binary-minimal binary-debug cli bench exporter-minimal verify-config test
```
[the remark this answers is not in the document]
Comment (chofnar) [reply .github/workflows/ci-cd.yml]: For 1: Moved logic of test-clean target to the specific Go test that needed it
For 2: Answered in a previous comment, changed to pattern "coverage*.txt"
For 3: Kept line unchanged

File: Makefile [hunk 6]
```diff
@@ -66,7 +76,7 @@ run-bench: binary bench
 
 .PHONY: test-clean
 test-clean:
-	$(shell sudo rm -rf /etc/containers/certs.d/127.0.0.1:8089/)
+	$(shell rm -rf /etc/containers/certs.d/127.0.0.1:8089/)
```
[the remark this answers is not in the document]
Comment (chofnar) [reply Makefile]: Moved logic to the Go test that needed it, removed the test-clean target

File: pkg/cli/client_sudo_test.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	"testing"
+	"time"
+
+	. "github.com/smartystreets/goconvey/convey"
+	"gopkg.in/resty.v1"
+	"zotregistry.io/zot/pkg/api"
+	"zotregistry.io/zot/pkg/api/config"
+)
+
+func TestSudoTLSNewControllerPrivilegedCert(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (chofnar) [reply pkg/cli/client_sudo_test.go]: As specified in the other comments - this is the test that got the logic of the "certs.d" and "test-clean" make targets added

File: pkg/storage/storage_fs_sudo_test.go (hunk 1 above)
Comment (chofnar) [reply pkg/storage/storage_fs_sudo_test.go]: Removed sudo command from test, as it now executes under "root" user (make being called with sudo make privileged-test)

File: pkg/cli/client_test.go [hunk 8]
```diff
@@ -1,5 +1,5 @@
-//go:build extended
-// +build extended
+//go:build extended || needprivileges
```
Comment (andaaron) [inline pkg/cli/client_test.go:1]: Why need this change?

**Review by andaaron — APPROVED** [review]

File: pkg/cli/client_test.go (hunk 8 above)
Comment (chofnar) [reply pkg/cli/client_test.go]: One of the tests with elevated privileges needed this before due to imports, but another change along the way must have fixed it. Removed it in the latest commit

**Review by andaaron — APPROVED** [review]

File: .github/workflows/ci-cd.yml [hunk 9]
```diff
@@ -59,8 +59,10 @@ jobs:
           cd $GITHUB_WORKSPACE
           if [[ $OS == "linux" && $ARCH == "amd64" ]]; then
             make OS=$OS ARCH=$ARCH
+            sudo env "PATH=$PATH" make privileged-test
+            make covhtml
```
[the remark this answers is not in the document]
Comment (chofnar) [reply .github/workflows/ci-cd.yml]: Made it one line

File: pkg/cli/client_elevated_test.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+			imageCmd := NewImageCommand(new(searchService))
+			imageBuff := bytes.NewBufferString("")
+			imageCmd.SetOut(imageBuff)
+			imageCmd.SetErr(imageBuff)
+			imageCmd.SetArgs(args)
+			err := imageCmd.Execute()
+			So(err, ShouldBeNil)
+		})
+
+		cmd = exec.Command("rm", "-rf", "/etc/containers/certs.d/127.0.0.1:8089/")
```
[the remark this answers is not in the document]
Comment (chofnar) [reply pkg/cli/client_elevated_test.go]: Moved rm command to a defer statement right after creating the dir

File: Makefile [hunk 11]
```diff
@@ -47,27 +47,28 @@ exporter-minimal: modcheck
...
 	go test -tags dev,extended,containers_image_openpgp -v -trimpath -race -timeout 15m -cover -coverpkg ./... -coverprofile=coverage-dev-extended.txt -covermode=atomic ./pkg/test/... ./pkg/storage/... ./pkg/extensions/sync/... -run ^TestInject
 	go test -tags dev,minimal,containers_image_openpgp -v -trimpath -race -cover -coverpkg ./... -coverprofile=coverage-dev-minimal.txt -covermode=atomic ./pkg/test/... ./pkg/storage/... ./pkg/extensions/sync/... -run ^TestInject
 
+.PHONY: privileged-test
+privileged-test: check-skopeo test/data $(NOTATION)
+	go test -tags needprivileges,extended,containers_image_openpgp -v -trimpath -race -timeout 15m -cover -coverpkg ./... -coverprofile=coverage-dev-needprivileges.txt -covermode=atomic ./pkg/storage/... ./pkg/cli/... -run ^TestElevatedPrivileges
+
+test/data: check-skopeo
+	$(shell mkdir -p test/data; cd test/data; ../scripts/gen_certs.sh; cd ${TOP_LEVEL}; skopeo --insecure-policy copy -q docker://public.ecr.aws/t0x7q1g8/centos:7 oci:${TOP_LEVEL}/test/data/zot-test:0.0.1;skopeo --insecure-policy copy -q docker://public.ecr.aws/t0x7q1g8/centos:8 oci:${TOP_LEVEL}/test/data/zot-cve-test:0.0.1)
```
[the remark this answers is not in the document]
Comment (chofnar) [reply Makefile]: Defined TESTDATA := $(TOP_LEVEL)/test/data and replaced all test/data references with $(TESTDATA)

**Review by andaaron — DISMISSED** [review]

File: Makefile [hunk 12]
```diff
@@ -47,28 +48,29 @@ exporter-minimal: modcheck
...
 	go test -tags minimal,containers_image_openpgp -v -trimpath -race -cover -coverpkg ./... -coverprofile=coverage-minimal.txt -covermode=atomic ./...
 	# development-mode unit tests possibly using failure injection
 	go test -tags dev,extended,containers_image_openpgp -v -trimpath -race -timeout 15m -cover -coverpkg ./... -coverprofile=coverage-dev-extended.txt -covermode=atomic ./pkg/test/... ./pkg/api/... ./pkg/storage/... ./pkg/extensions/sync/... -run ^TestInject
 	go test -tags dev,minimal,containers_image_openpgp -v -trimpath -race -cover -coverpkg ./... -coverprofile=coverage-dev-minimal.txt -covermode=atomic ./pkg/test/... ./pkg/storage/... ./pkg/extensions/sync/... -run ^TestInject
 	go test -tags stress,extended,containers_image_openpgp -v -trimpath -race -timeout 15m ./pkg/cli/stress_test.go
 
+.PHONY: privileged-test
+privileged-test: check-skopeo $(TESTDATA) $(NOTATION)
+	go test -tags needprivileges,extended,containers_image_openpgp -v -trimpath -race -timeout 15m -cover -coverpkg ./... -coverprofile=coverage-dev-needprivileges.txt -covermode=atomic ./pkg/storage/... ./pkg/cli/... -run ^TestElevatedPrivileges
```
Comment (rchincha) [inline Makefile:61]: coverage-dev-needprivileges.txt
^ where is this being combined to contribute to coverage?

File: .github/workflows/ci-cd.yml [hunk 13]
```diff
@@ -62,6 +62,7 @@ jobs:
           cd $GITHUB_WORKSPACE
           if [[ $OS == "linux" && $ARCH == "amd64" ]]; then
             make OS=$OS ARCH=$ARCH
+            sudo env "PATH=$PATH" make privileged-test covhtml
```
Comment (rchincha) [inline .github/workflows/ci-cd.yml:65]: No need to call 'covhtml' target here, the coverage ci-cd will pick up an *.txt as coverage.

Comment (rchincha) [reply .github/workflows/ci-cd.yml:65]: And in any case, add this file to `covhtml` target in the Makefile just to be consistent.

Comment (chofnar) [reply .github/workflows/ci-cd.yml]: > No need to call 'covhtml' target here, the coverage ci-cd will pick up an *.txt as coverage.

Removed covhtml from that line. I had put it there as I didn't know whether the coverage ci-cd would pick up the new needprivilege cover profile too ( -coverprofile=coverage-dev-needprivileges.txt in the new make target) before the covhtml target was run. The modification mentioned below makes covhtml include that file too. Or perhaps I should instead put one of the existing cover profiles when running the privileged-test target, and not a new one?

> And in any case, add this file to `covhtml` target in the Makefile just to be consistent.

Please check line 87 in the Makefile. Is that option agreeable?

**Review by andaaron — APPROVED** [review]

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**